packages/core/contracts/financial-templates/demo/OptimisticDepositBox.sol

```diff
@@ -0,0 +1,290 @@
+    }
+
+    /**
+     * @notice Starts a withdrawal request that allows the sponsor to withdraw `denominatedCollateralAmount`
+     * from their position denominated in the quote asset of the price identifier, following a Optimistic
+     * Oracle price resolution.
+     * @dev The request will be pending for the duration of the liveness period and can be cancelled at any
+     * time. Only one withdrawal request can exist for the user.
+     * @param denominatedCollateralAmount the quote-asset denominated amount of collateral requested to
+     * withdraw.
+     */
+    function requestWithdrawal(uint256 denominatedCollateralAmount)
+        public
+        noPendingWithdrawal(msg.sender)
+        nonReentrant()
+    {
+        OptimisticDepositBoxData storage depositBoxData = depositBoxes[msg.sender];
+        require(denominatedCollateralAmount > 0, "Invalid collateral amount");
+
+        // Update the position data for the user.
+        depositBoxData.withdrawalRequestAmount = denominatedCollateralAmount;
+        depositBoxData.withdrawalRequestTimestamp = getCurrentTime();
+
+        emit RequestWithdrawal(msg.sender, denominatedCollateralAmount, depositBoxData.withdrawalRequestTimestamp);
+
+        // A price request is sent for the current timestamp.
+        _requestOraclePrice(depositBoxData.withdrawalRequestTimestamp);
+    }
+
+    /**
+     * @notice After a withdrawal request (i.e., by a call to `requestWithdrawal`) and optimistic oracle
+     * price resolution, withdraws `depositBoxData.withdrawalRequestAmount` of collateral currency
+     * denominated in the quote asset.
+     * @dev Might not withdraw the full requested amount in order to account for precision loss.
+     * @return amountWithdrawn The actual amount of collateral withdrawn.
+     */
+    function executeWithdrawal() external nonReentrant() returns (uint256 amountWithdrawn) {
+        OptimisticDepositBoxData storage depositBoxData = depositBoxes[msg.sender];
+        require(
+            depositBoxData.withdrawalRequestTimestamp != 0 &&
+                depositBoxData.withdrawalRequestTimestamp <= getCurrentTime(),
+            "Invalid withdraw request"
+        );
+
+        // Get the resolved price or revert.
+        uint256 exchangeRate = _getOraclePrice(depositBoxData.withdrawalRequestTimestamp);
+
+        // Calculate denomated amount of collateral based on resolved exchange rate.
+        // Example 1: User wants to withdraw $1000 of ETH, exchange rate is $2000/ETH, therefore user to receive 0.5 ETH.
+        // Example 2: User wants to withdraw $2500 of ETH, exchange rate is $2000/ETH, therefore user to receive 1.25 ETH.
+        uint256 denominatedAmountToWithdraw = depositBoxData.withdrawalRequestAmount.div(exchangeRate);
+
+        // If withdrawal request amount is > collateral, then withdraw the full collateral amount.
+        if (denominatedAmountToWithdraw > depositBoxData.collateral) {
```

Comment:
Should this be >=? Otherwise, I think if the withdrawal amount is equal to the collateral, the collateral will be 0 after, but we will have never emitted the Ended... event.

Reply:
Good point; fixing.